sourced/ml/algorithms/uast_id_distance.py

```diff
@@ -0,0 +1,134 @@
+from collections import namedtuple
+from typing import Iterable, Tuple, Union
+from itertools import combinations
+
+import bblfsh
+
+from sourced.ml.algorithms.uast_ids_to_bag import UastIds2Bag
+from sourced.ml.utils import bblfsh_roles
+
+
+Point = namedtuple("Point", ["token", "info"])
+
+
+class Uast2IdDistance(UastIds2Bag):
+    """
+    Converts a UAST to a list of identifiers pair and UAST distance between.
+    distance should be defined in inheritors class
+
+    Be careful, __call__ is overridden here and return list instead of bag-of-words (dist).
+    """
+
+    DEFAULT_MAX_DISTANCE = 10  # to avoid collecting all distances we skip too big ones
+
+    def __init__(self, token2index=None, token_parser=None, max_distance=DEFAULT_MAX_DISTANCE):
+        """
+        :param token2index: The mapping from tokens to token key. If None, no mapping is performed.
+        :param token_parser: Specify token parser if you want to use a custom one. \
+            :class:'TokenParser' is used if it is not specified.
+        """
+        super().__init__(token2index=token2index, token_parser=token_parser)
+        self.max_distance = max_distance
+
+    def __call__(self, uast: bblfsh.Node) -> Iterable[Tuple[str, str, int]]:
+        """
+        Converts a UAST to a list of identifiers pair and UAST distance between.
+        The tokens are preprocessed by _token_parser.
+
+        :param uast: The UAST root node.
+        :return: a list of (from identifier, to identifier) and distance pairs.
+        """
+        for point1, point2 in combinations(self._process_uast(uast), 2):
+            if point1.token == point2.token:
+                continue  # We do not want to calculate distance between the same identifiers
+            distance = self.distance(point1, point2)
+            if distance < self.max_distance:
+                yield ((point1.token, point2.token), distance)
+
+    def distance(self, point1: Point, point2: Point) -> Union[int, float]:
+        """
+        Calculate distance between two points. A point can be anything. self._process_uast returns
+        list of points in the specific class.
+
+        :return: Distance between two points.
+        """
+        raise NotImplemented
+
+    def _process_uast(self, node: bblfsh.Node) -> Iterable[Point]:
+        """
+        Converts uast to points list. A point can be anything you need to calculate distance.
+        """
+        raise NotImplemented
```

Comment:
Are you sure this is needed? To me can be implemented in-place. Also I see copy-pasted code implementing this method below.

Reply:
wow, definitely it is needed.
there are different `_process_uast` functions. We can generalize it, but not now.

Reply:
I cannot pass such duplicated code, sorry, you need to refactor now.